From 1ab561c27d49c31117dbff2b307cb7feb56717d9 Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Thu, 7 Nov 2019 19:59:40 -0800
Subject: [PATCH 0682/1852] v4l: xilinx: xcsi2rxss: Source pad has same format
 as sink pad

commit 95ac2ca24d1bdf7ac0dbe57b9989e7ba1616a529 from
https://github.com/Xilinx/linux-xlnx.git

This IP doesn't do any color conversion. So sink and source pad should
have same color format. The format set on sink pad is set for source
pad.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-csi2rxss.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-csi2rxss.c b/drivers/media/platform/xilinx/xilinx-csi2rxss.c
index 5e2a241a8069..62e7336ec7e5 100644
--- a/drivers/media/platform/xilinx/xilinx-csi2rxss.c
+++ b/drivers/media/platform/xilinx/xilinx-csi2rxss.c
@@ -501,7 +501,7 @@ struct xcsi2rxss_state {
 	struct xcsi2rxss_core core;
 	struct v4l2_subdev subdev;
 	struct v4l2_ctrl_handler ctrl_handler;
-	struct v4l2_mbus_framefmt formats[2];
+	struct v4l2_mbus_framefmt formats;
 	struct v4l2_mbus_framefmt default_format;
 	const struct xvip_video_format *vip_format;
 	struct v4l2_event event;
@@ -1293,7 +1293,7 @@ __xcsi2rxss_get_pad_format(struct xcsi2rxss_state *xcsi2rxss,
 	case V4L2_SUBDEV_FORMAT_TRY:
 		return v4l2_subdev_get_try_format(&xcsi2rxss->subdev, cfg, pad);
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xcsi2rxss->formats[pad];
+		return &xcsi2rxss->formats;
 	default:
 		return NULL;
 	}
@@ -1358,6 +1358,15 @@ static int xcsi2rxss_set_format(struct v4l2_subdev *sd,
 	__format = __xcsi2rxss_get_pad_format(xcsi2rxss, cfg,
 						fmt->pad, fmt->which);
 
+	/*
+	 * If trying to set format on source pad, then
+	 * return the format set on sink pad
+	 */
+	if (fmt->pad == 0) {
+		fmt->format = *__format;
+		goto unlock_set_fmt;
+	}
+
 	/* Save the pad format code */
 	code = __format->code;
 
@@ -1400,6 +1409,7 @@ static int xcsi2rxss_set_format(struct v4l2_subdev *sd,
 		__format->height = fmt->format.height;
 	}
 
+unlock_set_fmt:
 	mutex_unlock(&xcsi2rxss->lock);
 
 	return 0;
@@ -1914,8 +1924,7 @@ static int xcsi2rxss_probe(struct platform_device *pdev)
 	xcsi2rxss->default_format.width = XCSI_DEFAULT_WIDTH;
 	xcsi2rxss->default_format.height = XCSI_DEFAULT_HEIGHT;
 
-	xcsi2rxss->formats[0] = xcsi2rxss->default_format;
-	xcsi2rxss->formats[1] = xcsi2rxss->default_format;
+	xcsi2rxss->formats = xcsi2rxss->default_format;
 
 	/* Initialize V4L2 subdevice and media entity */
 	subdev = &xcsi2rxss->subdev;
-- 
2.31.1

